### PR TITLE
[1.11] Update BuildCraft API to maven dependency

### LINKED ISSUE #4513: Crash with newest buildcraft version
#### Issue description:
Crash while loading with exception:
net.minecraftforge.fml.common.LoaderExceptionModCrash: Caught exception from MekanismGenerators (mekanismgenerators)
Caused by: java.lang.NoSuchMethodError: buildcraft.api.fuels.IFuel.getFluid()Lnet/minecraftforge/fluids/Fluid;

#### Steps to reproduce:
1. Install [latest Buildcraft version](https://www.mod-buildcraft.com/buildcraft-7995-alpha.html) 7.99.5 + Mekanism 9.3.2.307
2. Start game.
3. ???
4. Profit!

#### Version:
**Forge:** 13.20.1.2393
**Mekanism:** 9.3.2.307
**Other relevant version:** BuildCraft 7.99.5 Alpha

#### link here:

[Crash](http://paste.ubuntu.com/25035842/)

### PR DESCRIPTION
**Pull request by thommy101**
BuildCraft API changed with some breaking changes. This commit updates the used API to compile with and updates FuelHandler & MekanismGenerators to use this new API.

BC's API currently stores the energy value of fuels as Jules. If I'm correct, Mekanism uses Jules internally, so no conversion from RF needed.

Fixes #4513

--EDIT--
Before merging, please check if the values are correct!

Current values:
Ethene gives 240J/tick (600 RF) as value to BC
Conversion of BC Joules is 1:1 with Mekanism Joules (BC switched from RF to J as value to return from their API, with their RF/J calculation as follows: `power = rf * MjApi.MJ / 20` where `MjAPI.MJ` equals 1.000.000L)

To stay in sync with BC there needs to be a conversion.
1 BCJ (not micro, just Joules) being 20 RF, which is 8 Mekanism Joules.

### CODE REVIEW

**Comment by AlexIIL** [comment]
I'd recommend either using a git submodule, or depending on the BBC maven https://mod-buildcraft.com/maven/com/mod-buildcraft/buildcraft-api/ and using version 7.99.6

**Comment by thommy101** [comment]
I just read your message with the BC Maven repo, updating locally, some testing and I'll update this ;)

File: src/main/java/mekanism/common/FuelHandler.java [hunk 1]
```diff
@@ -56,7 +58,7 @@ public FuelGas(int duration, double energyDensity)
 		public FuelGas(IFuel bcFuel)
 		{
 			burnTicks = bcFuel.getTotalBurningTime() / Fluid.BUCKET_VOLUME;
-			energyPerTick = bcFuel.getPowerPerCycle() * general.FROM_RF;
+			energyPerTick = bcFuel.getPowerPerCycle() * MjAPI.MJ;
```
Comment (thommy101) [inline src/main/java/mekanism/common/FuelHandler.java]: Thanks for looking at this!

What I see from this, is a cross mod compatability feature allowing BC fuel burn in the Mekanism GasGenerator. It's about adding features to Mekanism, not to BC ;)

The desired result is energyPerTick in Joules as that's what Mekanism uses internally if I'm right. 
With 1J being: 10EU, 2.5RF (and Tesla, ForgeEnergy).

If I'm reading the JavaDoc right, getPowerPerCycle returns micro mj, so multiplying with MjAPI.MJ (1.000.000) should return Joule right?

File: src/main/java/mekanism/generators/common/MekanismGenerators.java [hunk 2]
```diff
@@ -101,13 +101,13 @@ public void postInit(FMLPostInitializationEvent event)
...
-				if(!(s.getFluid() == null || GasRegistry.containsGas(s.getFluid().getName())))
+				if(!(s.getFluid() == null || GasRegistry.containsGas(s.getFluid().getFluid().getName())))
 				{
-					GasRegistry.register(new Gas(s.getFluid()));
+					GasRegistry.register(new Gas(s.getFluid().getFluid()));
 				}
 			}
 
-			BuildcraftFuelRegistry.fuel.addFuel(MekanismFluids.Ethene.getFluid(), (int)(240 * general.TO_RF), 40 * Fluid.BUCKET_VOLUME);
+			BuildcraftFuelRegistry.fuel.addFuel(MekanismFluids.Ethene.getFluid(), 240, 40 * Fluid.BUCKET_VOLUME);
```
Comment (thommy101) [inline src/main/java/mekanism/generators/common/MekanismGenerators.java]: I don't know the intentions with the 240 (600 RF). If I'm reading this right, it's about the generated power per tick. In that case I may think 600RF/t is the right value.

But only @aidancbrady could confirm/deny this value.

File: src/main/java/mekanism/common/FuelHandler.java (hunk 1 above)
Comment (AlexIIL) [inline src/main/java/mekanism/common/FuelHandler.java]: I'm not familiar with mekanism's joule system, however getPowerPerCycle will return something like "6000000" (6 * MjApi.MJ). Currently the conversion from RF to MJ is `power = rf * MjApi.MJ / 20`, so I don't think you can convert exactly if 1J = 2.5 RF

Comment (thommy101) [inline src/main/java/mekanism/common/FuelHandler.java]: Uhm.. I did read the value the wrong way, it should be a division of course. The conversion between Mekanism Joules and BC Joules, I don't know about it...

It's updated, I'll leave the confirmation of the values to Aidan to confirm ;)
Thanks for the input! If you see some more stuff, please let me know!

Comment (AlexIIL) [inline src/main/java/mekanism/common/FuelHandler.java]: Sure :)

Although if this is all the bc integration Mekanism has then there's nothing else to comment on :)

Comment (thommy101) [inline src/main/java/mekanism/common/FuelHandler.java]: Nothing to comment on the values of BC ofcourse, but the conversion to Mekanism values might be wrong. I'll update the first post with the points that needs some checking.

To make things clear, nothing wrong with BC, my conversion of BC values to Mekanism values might be messed up as I don't know exactly what kind off values Mekanism uses internally.

Comment (AlexIIL) [inline src/main/java/mekanism/common/FuelHandler.java]: Err woops, it's actually divided by 10, not 20.

Comment (AlexIIL) [inline src/main/java/mekanism/common/FuelHandler.java]: Although I'm not sure that will be the final conversion value...

Comment (thiakil) [inline src/main/java/mekanism/common/FuelHandler.java]: Is there a way to get the BCmj -> RF rate from the API? I believe Mek has separate conversion config values, but we could at least use that to generate a default value.

**Comment by aidancbrady** [comment]
Going to hold off merging until I have some free time to find some solutions. Hopefully I'll be able to get some real work done towards the end of this coming weekend.

File: src/main/java/mekanism/generators/common/MekanismGenerators.java [hunk 3]
```diff
@@ -101,13 +102,13 @@ public void postInit(FMLPostInitializationEvent event)
 		{
 			for(IFuel s : BuildcraftFuelRegistry.fuel.getFuels())
 			{
-				if(!(s.getFluid() == null || GasRegistry.containsGas(s.getFluid().getName())))
+				if(!(s.getFluid() == null || GasRegistry.containsGas(s.getFluid().getFluid().getName())))
```
Comment (thiakil) [inline src/main/java/mekanism/generators/common/MekanismGenerators.java]: Purely stylistic here (and I know it's not originally your code Tom), but should this be changed to `s.getFluid != null && !GasRegistry.contains...` ?

Seems a bit confusing as-is @aidancbrady

File: src/main/java/mekanism/generators/common/MekanismGenerators.java [hunk 4]
```diff
@@ -102,13 +102,13 @@ public void postInit(FMLPostInitializationEvent event)
...
 			{
-				if(!(s.getFluid() == null || GasRegistry.containsGas(s.getFluid().getFluid().getName())))
+				if(s.getFluid() != null && !GasRegistry.containsGas(s.getFluid().getFluid().getName()))
 				{
 					GasRegistry.register(new Gas(s.getFluid().getFluid()));
 				}
 			}
 
-			BuildcraftFuelRegistry.fuel.addFuel(MekanismFluids.Ethene.getFluid(), 240 * MjAPI.MJ, 40 * Fluid.BUCKET_VOLUME);
+			BuildcraftFuelRegistry.fuel.addFuel(MekanismFluids.Ethene.getFluid(), (int)(240 * general.TO_RF / 20 * MjAPI.MJ), 40 * Fluid.BUCKET_VOLUME);
```
Comment (AlexIIL) [inline src/main/java/mekanism/generators/common/MekanismGenerators.java]: You should cast to long, not int. Although the numbers might be to low to matter.

**Comment by thommy101** [comment]
Changed the conversion rate between BuildCraft-RF-Mekanism.
Currently using the following values:
1.000.000 Micro BuildCraft MJ = 1 BuildCraft MJ = 20 RF = 8 MekanismJoules.

The value 1.000.000 is defined as MjAPI.MJ
The value 20 is hardcoded (one in FuelHandler, one in MekanismGenerators)
The RF/Mekanim Joule is config defined (MekanismConfig.general.FROM_RF and TO_RF)

Tested without BC running and it's not crashing with class not found exceptions.

File: src/main/java/mekanism/generators/common/MekanismGenerators.java (hunk 4 above)
Comment (thommy101) [inline src/main/java/mekanism/generators/common/MekanismGenerators.java]: Don't know why my IDE suggested integer, changing to long as that's what the API defines (value could get > MAX_INT as general.TO_RF is user defined)

**Comment by aidancbrady** [comment]
I was originally planning on adding full integration for BC's changes, but I'm getting pretty tired of integrating with newly-spawned energy frameworks. If this one catches on I'll follow through, but in the meantime I'll leave it to the current integration (Tesla, Forge Energy, RF, IC2, Mekanism)

**Comment by Ricket** [comment]
FYI. After pulling this change, I am having difficulty building:

```
FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring root project 'Mekanism'.
> Could not resolve all dependencies for configuration ':deobfCompile'.
   > Could not resolve com.mod-buildcraft:buildcraft-api:7.99.6.
     Required by:
         :Mekanism:1.11.2-9.3.3.homebaked
      > Could not resolve com.mod-buildcraft:buildcraft-api:7.99.6.
         > Could not get resource 'https://mod-buildcraft.com/maven/com/mod-buildcraft/buildcraft-api/7.99.6/buildcraft-api-7.99.6.pom'.
            > Could not GET 'https://mod-buildcraft.com/maven/com/mod-buildcraft/buildcraft-api/7.99.6/buildcraft-api-7.99.6.pom'.
               > peer not authenticated

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED
```

Have not found the solution yet. When I load the url in my browser, the cert checks out, it's a LetsEncrypt cert.

**Comment by AlexIIL** [comment]
@Ricket previous people have been able to update java to a newer version to fix it - I must have accidentally selected a newer protocol or something like that (when configuring the cert), which previous versions of the JVM can't use.

**Pull request merged**